MH-13129 More configuration options for thumbnails

staubesv opened the pull request:
- Don't hardcoded channel ID of configurable publication channel
- Allow multiple encoding profiles to be set for both OAI-PMH and configurable publication channel
- Allow flavor and tags to be set individually for both OAI-PMH and configurable publication channel
- Use special encoding profile for the extraction of the master image
- Optimize preview extraction in case auto distribution is disabled by having a special encoding profile for that case
- Allow auto distribution to be turned of for individual publication channels
- Extended convertImageSync to support multiple encoding profiles

This PR is based on work of Plapadoo and replaces #452.

staubesv committed:
MH-13129 More configuration options for thumbnails

- Don't hardcoded channel ID of configurable publication channel
- Allow multiple encoding profiles to be set for both OAI-PMH and configurable publication channel
- Allow flavor and tags to be set individually for both OAI-PMH and configurable publication channel
- Use special encoding profile for the extraction of the master image
- Optimize preview extraction in case auto distribution is disabled by having a special encoding profile for that case
- Allow auto distribution to be turned of for individual publication channels
- Extended convertImageSync to support multiple encoding profiles

staubesv commented:
@lkiesow & @mtneug I would prefer to do this work against 6.x as it changes code that is just introduced with 6.0. While it does more than a simple bugfix, the code it touches is only relevant for 6.0 ("thumbnail feature") and avoids 6.0 users to have to adopt their configuration for this new feature in 7.0 again.

Please let me know in case you want me to redirect this to develop.

mtneug commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/AdminUIConfiguration.java:
```diff
@@ -44,19 +85,27 @@
+  public static final String OPT_THUMBNAIL_DISTRIBUTION_CONFIGURABLE_FLAVOR = "thumbnail.distribution.oaipmh.flavor";
+  public static final String OPT_THUMBNAIL_DISTRIBUTION_CONFIGURABLE_TAGS = "thumbnail.distribution.oaipmh.tags";
+  public static final String OPT_THUMBNAIL_DISTRIBUTION_CONFIGURABLE_PROFILES =
+      "thumbnail.distribution.oaipmh.profiles";
```
Copy paste error `oaipmh -> configurable`?

staubesv replied:
Fixed.

mtneug commented:
@staubesv after skimming through the PR, I'm undecided if this should go into 6.x. I know that breaking configuration changes are painful for maintainers (guilty myself 😐), but these changes result from new features and are not security related. Also, breaking changes between major versions are normal. Basically the question is, should PRs, which improve upon features that are not yet released, be accepted into release branches. Is this change really so special to allow that?

staubesv committed:
MH-13129 More configuration options for thumbnails

- Don't hardcoded channel ID of configurable publication channel
- Allow multiple encoding profiles to be set for both OAI-PMH and configurable publication channel
- Allow flavor and tags to be set individually for both OAI-PMH and configurable publication channel
- Use special encoding profile for the extraction of the master image
- Optimize preview extraction in case auto distribution is disabled by having a special encoding profile for that case
- Allow auto distribution to be turned of for individual publication channels
- Extended convertImageSync to support multiple encoding profiles

staubesv committed:
Merge branch 't/MH-13129-more-thumbnail-options' of https://github.com/switch-ch/opencast into t/MH-13129-more-thumbnail-options

# Conflicts:
#	modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/endpoint/ComposerRestService.java

staubesv committed:
MH-13129 Fix names of constants

staubesv commented:
@mtneug The original motivation to raise this against 6.x rather than develop was that it does not add a feature but rather improves a feature that is being added to 6.x. Meanwhile, though, the 6.x release is close and it is admittedly a bit late to add this.

I've redirected this PR to develop.

staubesv committed:
MH-13129 Fix merge errors

KatrinIhler commented on docs/guides/admin/docs/configuration/admin-ui/thumbnails.md:
```diff
@@ -35,12 +35,19 @@ attachment with the flavor as specified in the Admin UI configuration:
+
+In case that automatic distribution is enabled, a master thumbnail image that will be converted into all of the
+formats required is extracted and the thumbnail preview image will be based on that master image.
+The following encoding profile is then used to downscale the master image:
```
This section confuses me. So if auto-distribution is on, the image is extracted using editor-thumbnail-preview and then downscaled, and if auto-distribution is off... is editor.thumbnail.preview.downscale used to directly extract the preview with no image in between?

staubesv replied:
I've improved the explanation of the thumbnail preview. Hope that things are more clear now.

KatrinIhler commented on docs/guides/admin/docs/configuration/admin-ui/thumbnails.md:
```diff
@@ -131,33 +137,61 @@ To avoid the situation that a user needs to start a workflow just to update the
-Currently, this is supported for External API publication channels (created by WOH publish-configure) and OAI-PMH
-publication channels (created by WOH publish-oaipmh).
+Currently, this is supported for configurable publication channels (created by WOH publish-configure) as used by the
+External API and OAI-PMH publication channels (created by WOH publish-oaipmh).
```
This confuses me as well. So publish-oaipmh is for OAIPMH and publish-configure is...for the external API? Everything else? Both?

staubesv replied:
The sentence quite clearly states "configurable publications channels [...] as used by the External API" and "OAI-PMH publication channels". Any suggestion of how to improve this?

KatrinIhler replied:
Maybe add a , after External API to make it clearer where the second half of the sentence belongs to?

staubesv replied:
Fixed.

KatrinIhler commented on docs/guides/admin/docs/configuration/admin-ui/thumbnails.md:
```diff
@@ -131,33 +137,61 @@ To avoid the situation that a user needs to start a workflow just to update the
 
-The following encoding profile is used for thumbnail extraction (default thumbnail and snapshot thumbnail):
+    # Default: editor.thumbnail.master
+    #thumbnail.master.profile=editor.thumbnail.master
```
This confuses me further (mainly in conjunction with the preview docs above). I think if this is the profile for the master image, I mainly have a problem with understanding the difference between editor.thumbnail.preview and editor.thumbnail.preview.downscale and their different uses, could you elaborate?

staubesv replied:
The new explanation of the thumbnail preview should make this clear.

KatrinIhler commented on docs/guides/admin/docs/configuration/admin-ui/thumbnails.md:
```diff
@@ -131,33 +137,61 @@ To avoid the situation that a user needs to start a workflow just to update the
+
+    # The ID of the configurable publication channel
+    # Default: api
+    #thumbnail.distribution.configurable.channel=api
```
I expected _oaipmh_ or something here - so is this OAIPMH specific or applicable for different publication channels?

staubesv replied:
Don't exactly understand the question...?

KatrinIhler commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/AdminUIConfiguration.java:
```diff
@@ -240,26 +292,26 @@ public void updated(Dictionary<String, ?> properties) throws ConfigurationExcept
+    // Encoding profile used to extract the thumbnail preview image
+    thumbnailPreviewProfile = StringUtils.defaultString(
+      (String) properties.get(OPT_THUMBNAIL_PREVIEW_PROFILE), DEFAULT_THUMBNAIL_PREVIEW_PROFILE);
+    logger.debug("Thumbnail preview encoding profile set to '{}'", thumbnailPreviewProfile);
```
ThumbnailDistributionSettings doesn't have a custom toString(), so I think it would just log a rather useless standard object description here, no?

staubesv replied:
thumbnailPreviewProfile is a Java string.

KatrinIhler commented on modules/composer-service-api/src/main/java/org/opencastproject/composer/api/ComposerService.java:
```diff
@@ -248,20 +248,21 @@ Job image(Track sourceTrack, String profileId, Map<String, String> properties) t
-   * Synchronously converts the given image to a different image format using the specified image profile. Please note that
-   * synchronously doing this means, that the workload cannot be distributed amongst all nodes.
+   * Synchronously converts the given image to different image formats using the specified encoding profiles. Please
+   * note that synchronously doing this means, that the workload cannot be distributed amongst all nodes.
```
no ,

staubesv replied:
Fixed.

KatrinIhler commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/ThumbnailImpl.java:
```diff
@@ -371,38 +394,44 @@ private URI updateOaiPmh(final MediaPackage mp, final MediaPackageElementFlavor
+    return downscaled.stream().map(a -> cloneAttachment(attachment, a.getURI())).collect(Collectors.toList());
   }
 
   private URI updateExternalPublication(final MediaPackage mp, final MediaPackageElementFlavor trackFlavor)
```
Maybe rename this to updateConfigurablePublication?

staubesv replied:
Fixed.

KatrinIhler commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/ThumbnailImpl.java:
```diff
@@ -371,38 +394,44 @@ private URI updateOaiPmh(final MediaPackage mp, final MediaPackageElementFlavor
-    final Attachment downscaled = composerService.convertImageSync(attachment, conversionProfile);
-    attachment.setURI(downscaled.getURI());
+    final List<Attachment> downscaled = composerService.convertImageSync(attachment, conversionProfiles);
+    return downscaled.stream().map(a -> cloneAttachment(attachment, a.getURI())).collect(Collectors.toList());
```
Why do we need to clone this here?

staubesv replied:
The comment just above those lines describes this: The attachment returned by the composer service does not have the correct tags, flavor, etc.

KatrinIhler commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/ThumbnailImpl.java:
```diff
@@ -301,31 +309,35 @@ private void archive(final MediaPackage mp) {
 
     // We have to update the external publication to contain the new thumbnail as an attachment
-    final Optional<Publication> externalPublicationOpt = getPublication(mp, "api");
+    final Optional<Publication> externalPublicationOpt = getPublication(mp, distributionConfigurable.getChannelId());
```
Shouldn't this be distributionOAIPMH?

staubesv replied:
For whatever reason the OAI-PMH records do contain the publication elements. Since changing the thumbnail potentially implies a change for the publication element of "the" configurable publication channel, this needs to be updated, too.

KatrinIhler reviewed (commented):
First pass. Was a bit confused, so lots of questions. Another one: How does updating the configurable channel differ from updating OAIPMH? Couldn't OAIPMH maybe be just another configurable channel and we allow multiple configurable channels?

staubesv committed:
MH-13129 Code Review

- Better explanation of how the thumbnail preview is generated
- Rename externalPublication to configurablePublication for sake of consistency
- Fixed typos

staubesv commented:
Hi @KatrinIhler 

> First pass. Was a bit confused, so lots of questions. Another one: How does updating the configurable channel differ from updating OAIPMH?

It is basically a different kind of publication channel which requires different handling.

> Couldn't OAIPMH maybe be just another configurable channel and we allow multiple configurable channels?

No, that does not work. The publication elements of configurable publication channels are different from those of other publication channels and need to be created in a different way.

I think I resolve your comments.

-> Your turn again

staubesv committed:
MH-13129 Code Review

- Remove trailing whitespace

staubesv committed:
MH-13129 Code Review

- Fix mvn site

staubesv committed:
MH-13129 Code Review

Improve description of what kinds of publication channels support automatic distribution

staubesv commented:
@KatrinIhler As far as I see, there are no comments left for me to resolve. Is this correct?

KatrinIhler commented:
@staubesv You're right. I'm planning to take another look at this, probably tomorrow.

KatrinIhler reviewed (approved):
Aaand then it's February. I'm so sorry this took so long, this became like my nemesis among the pull requests... But finally tested and approved!
